### ISSUE TITLE
Meteor/Iron Router is reloading entire page when changing Routes

### ISSUE THREAD

**Issue by ashah888** [iron-meteor/iron-router#1149]
I'm noticing that every time Meteor/Iron Router changes routes, the entire page gets refreshed.

Why is this happening and how can I prevent it?

The way I have routes setup is something like this for every new route...

```
Router.route('/route/:_some_param', {
    name: 'something',
    onBeforeAction: function(){ // Something },
    waitOn: function(){ // Some subscription },
    action: function(){ // Something },
})
```

I have several routes setup this way and whenever I move from page to page that require different routes, Meteor reloads the entire page. However when I stay on one route and just change params, Meteor only reloads what's necessary, which is what I want all the time.

Is my code/logic correct?

**Comment by jordangarside** [comment]
I second this issue, I'll make a repo sometime soon. My layout is getting rendered on every page load. If I have a waitOn specified it will get triggered twice for 1 route. Somehow it's not triggering Template.layout.rendered multiple times, but it is triggering an angular controller specified inside the  layout template each route.

**Comment by benmgreene** [comment]
I've run into situations in which, after navigating to a different IR route, the browser will then reload the entire page. I don't mean reloading the entire layout, I mean actually downloading the entire app from the server again.

I'm able to reproduce this behavior by navigating to a route which:

1. Has it's path as `/`
2. Has a `subscriptions` function that returns `[]`

I have no clue why this happens, but it ain't good.

**Comment by saeho** [comment]
For those with this issue:

I used to run into issues very similar to this, semi-frequently.
Eventually I discovered that this can be caused by many different cases of faulty code.

In my case, I discovered that adding events to $(window) or $(document) that interfered with Meteor's own events caused Meteor to refresh. I can't remember exactly what the event was but events such as "scroll" events come to mind... It's safe to do scroll events but the way I did it before interfered with one of Meteor's events.

I also remember that doing $(window).off() caused Meteor to refresh, most likely because Meteor probably has its own $(window) events and doing $(window).off() probably turned them off.

In conclusion I was able to fix multiple cases of this issue by properly turning off events, being careful about how I wrote my event handler codes and making sure there are no faulty codes.

I think it would be helpful if Meteor logged some sort of error before the refresh happens.

**Comment by kevinpeter** [comment]
In my situation the problem was, that there was a click-listener on a parent element. In the function of the click-listener was e.`stopPropagation();` called. When i remove it, it works well.

**Comment by rteslaru** [comment]
@kevinpeter that's exactly what I am experiencing with meteor 1.2.0.2 and iron-router 1.0.12 -- full page reload on `e.stopPropagation()` on a click listener. Did you figure out a workaround?

**Comment by kevinpeter** [comment]
The solution for me, was to remove the code line `e.stopPropagation();`
If I included this code lines, it didn't wordk well. I think the main problem is, that iron-router has a handler for all <a>-Tag clicks. And if you call `e.stopPropagation();` then your browser didn't call the handler from iron router, so the browser, make an origin reload.

**Comment by rteslaru** [comment]
Same here, had to remove `e.stopPropagation();` to get it to work properly. Isn't this an iron-router bug?

**Comment by greggman** [comment]
I'm looking for a solution as well. Mine is only tangentially related to a click. I basically have

     "click #save": function() {
        Meteor.call("save", data, function(err, result) {
           if (!err) {
             Router.go(result);
        });
      });

It can take several seconds to save over a slow connection so the callback from "save" doesn't come back for seconds. When it does I want to change the location to the URL for the new document but meteor is also reloading the page. It works. But since my site has music that's not ideal.

The URL is something like "/doc/<id>" as in "/doc/rfkeifj39fu43j9fi" and there is a "subscriptions" section for that route. So maybe like mentioned above those are part of the triggers

**Comment by entioentio** [comment]
@saeho $(window).trigger('resize') is definetely one of the events :)

**Comment by realfresh** [comment]
Just bumped into this problem, for me what the issue ended up being was attaching events to the $(document). Once I changed that to $(window), my issues were resolved. I also found that attaching $(document) events in the on-rendered didnt end up running properly, I had to use $(window).

I am using stopPropagation() else where and it isnt causing any issues for me

**Comment by heberallred** [comment]
I was running into this same issue with:
`event.stopPropagation();` 
inside a Blaze event (for an "a" hyperlink tag) causing the page to fully reload.

I have a tree structure of nested DIV elements so I need propagation to be stopped or it will trigger for all parent items as well.  Here's a workaround I have done which should work in the case of nested elements (and does not break Iron Router page causing a page reload).  Please let me know if you have any better work-arounds:
```
let lastEventTime = new Date().getTime();  // Above in the file, maybe at the top

'click .recordItem': function (event, instance) {
  // Manual stop propagation of duplicate events that trigger within less than 20 milliseconds
  if (new Date().getTime() - lastEventTime < 20) return;
  lastEventTime = new Date().getTime();

  // Normal event code here
}
```

**Comment by MohammedEssehemy** [comment]
for me the problem was adding target attribute on the anchor tag, which caused full page refresh even if the value is `_self`.
